### PR TITLE
Set up mypy and fix mypy errors (non-strict)

### PR DESCRIPTION
**Pull request by eggplants**
image

I have set up mypy and fixed mypy errors.

---

To set up mypy:

- Added `mypy` to test dependencies and execute in test CI
  - Ref: https://github.com/eggplants/tweepy/actions/runs/1950162070
- Added `mypy` configuration to `setup.cfg`
  - Please uncomment `strict = True` if you want to make `mypy` more strict

---

To fix mypy errors:

- Added `types-oauthlib` and `types-requests` to dependencies
- Ignored types of `requests_oauthlib` and `bson` since they are untyped and any stubs aren't provided
  - Ref: https://github.com/requests/requests-oauthlib/issues/428
  - Ref: https://github.com/py-bson/bson/issues/114
- Excluded `None` of `self.access_token` from `Client`'s some methods
  - Some class methods of `Client` raise `TypeError` if `self.access_token` is `None` before `self.access_token.partition("-")`

### CODE REVIEW

**Comment by eggplants** [comment]
TIPS: `mypy --strict tweepy` reports `Found 909 errors in 20 files (checked 22 source files)`.

**Review by Harmon758 — CHANGES_REQUESTED** [review]
Regarding:
> - Excluded `None` of `self.access_token` from `Client`'s some methods
>   - Some class methods of `Client` raise `TypeError` if `self.access_token` is `None` before `self.access_token.partition("-")`

I had an implementation that adds support for OAuth 2.0 Authorization Code Flow with those methods in a stash (it was just missing documentation), and I've just pushed it with https://github.com/tweepy/tweepy/commit/0157d0b2abcfe40e4e5b77c3d8f733f67ebef9d9 and merged it into the `type-hints` branch.
I'd appreciate if you could resolve the conflicts now, and I'll give this a review afterwards.
Thanks for the PR, and sorry for the hassle.

**Comment by eggplants** [comment]
@Harmon758 ready for review.

File: tweepy/cache.py [hunk 1]
```diff
@@ -129,7 +130,7 @@ class FileCache(Cache):
     """File-based cache"""
 
     # locks used to make cache thread-safe
-    cache_locks = {}
+    cache_locks: Dict[str, threading.Lock] = {}
```
Comment (Harmon758) [inline tweepy/cache.py:133]: `dict` can be used here with `from __future__ import annotations`.

File: tweepy/client.py [hunk 2]
```diff
@@ -8,15 +8,15 @@
...
+    from functools import cache  # type: ignore[attr-defined]
 except ImportError:  # Remove when support for Python 3.8 is dropped
     from functools import lru_cache
     cache = lru_cache(maxsize=None)
 
 import logging
 from platform import python_version
 import time
-from typing import Any, TYPE_CHECKING
+from typing import Any, Iterable, TYPE_CHECKING
```
Comment (Harmon758) [inline tweepy/client.py:19]: This additional import should probably be in `if TYPE_CHECKING`.

File: tweepy/streaming.py [hunk 3]
```diff
@@ -964,6 +964,6 @@ class StreamRule(NamedTuple):
     .. _operators: https://developer.twitter.com/en/docs/twitter-api/tweets/search/integrate/build-a-query
     .. _Academic Research Project: https://developer.twitter.com/en/docs/projects
     """
-    value: str = None
-    tag: str = None
-    id: str = None
+    value: Optional[str] = None
+    tag: Optional[str] = None
+    id: Optional[str] = None
```
Comment (Harmon758) [inline tweepy/streaming.py:969]: `str | None` can be used with `from __future__ import annotations`.

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -18,7 +18,10 @@ jobs:
       - name: Install dependencies
         run: |
           python -m pip install --upgrade pip
-          pip install .[dev,test]
+          pip install .[dev,test,async]
+      - name: Run mypy
```
Comment (Harmon758) [inline .github/workflows/test.yml:22]:
```suggestion
      - name: Type-check
```

**Review by Harmon758 — CHANGES_REQUESTED** [review]

File: tweepy/client.py (hunk 2 above)
Comment (Harmon758) [reply tweepy/client.py]: I've resolved this in the `type-hints` branch.

File: setup.py [hunk 5]
```diff
@@ -41,7 +42,9 @@
         "oauthlib>=3.2.0,<4",
         "requests>=2.27.0,<3",
         "requests-oauthlib>=1.2.0,<2",
-        "typing-extensions>=4.0,<5"
+        "typing-extensions>=4.0,<5",
+        "types-oauthlib>=3.1.6,<4",
+        "types-requests>=2.27.11,<3",
```
[the remark this answers is not in the document]
Comment (Harmon758) [reply setup.py:47]: I've added a new `types` extra, so these should go there, and it needs to be installed in the GitHub Action workflow.

File: tweepy/cache.py (hunk 1 above)
Comment (eggplants) [reply tweepy/cache.py]: ed58b61

File: tweepy/streaming.py (hunk 3 above)
Comment (eggplants) [reply tweepy/streaming.py]: ed58b61

File: setup.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eggplants) [reply setup.py]: d4892ab

[the remark this answers is not in the document]
Comment (eggplants) [reply setup.py]: 39a2535

File: setup.py [hunk 6]
```diff
@@ -51,7 +52,11 @@
          ],
         "socks": ["requests[socks]>=2.27.0,<3"],
         "test": tests_require,
-        "types": ["typing-extensions>=4.0,<5"],
+        "types": [
+            "typing-extensions>=4.0,<5",
+            "types-oauthlib>=3.1.6,<4",
+            "types-requests>=2.27.11,<3",
```
Comment (Harmon758) [inline setup.py:58]:
```suggestion
            "types-oauthlib>=3.1.6,<4",
            "types-requests>=2.27.11,<3",
            "typing-extensions>=4.0,<5",
```

File: tweepy/cache.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 # Copyright 2009-2022 Joshua Roesslein
 # See LICENSE for details.
 
+from __future__ import annotations
```
Comment (Harmon758) [inline tweepy/cache.py:5]:
```suggestion
from __future__ import annotations

```

File: tweepy/streaming.py [hunk 8]
```diff
@@ -4,6 +4,7 @@
 
 # Appengine users: https://developers.google.com/appengine/docs/python/sockets/#making_httplib_use_sockets
 
+from __future__ import annotations
```
Comment (Harmon758) [inline tweepy/streaming.py:7]:
```suggestion
from __future__ import annotations

```

**Review by Harmon758 — CHANGES_REQUESTED** [review]

**Review by Harmon758 — APPROVED** [review]

**Pull request merged**